### PR TITLE
Update: Add un-downloadable error case

### PR DESCRIPTION
**Pull request by tonyjin**
Add error message for un-downloadable file and support for a link button instead of a download button in the error viewer.

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @tonyjin has signed the CLA. Thanks for the pull request!

File: src/lib/viewers/error/PreviewErrorViewer.js [hunk 1]
```diff
@@ -91,8 +94,11 @@ class PreviewErrorViewer extends BaseViewer {
         this.iconEl.innerHTML = this.icon;
         this.messageEl.textContent = displayMessage;
 
-        // Add optional download button
-        if (checkPermission(file, PERMISSION_DOWNLOAD) && showDownload && Browser.canDownload()) {
+        // Add optional link or download button
+        const { linkMsg, linkUrl } = err;
```
Comment (jeremypress) [inline src/lib/viewers/error/PreviewErrorViewer.js]: Where are we adding this linkUrl to the error object?

Comment (tonyjin) [reply src/lib/viewers/error/PreviewErrorViewer.js]: In a separate patch when the field is ready.

Comment (tonyjin) [reply src/lib/viewers/error/PreviewErrorViewer.js]: Ignore what I said, I've added the check for `is_download_available` into this patch now.

File: src/lib/Preview.js [hunk 2]
```diff
@@ -923,6 +924,16 @@ class Preview extends EventEmitter {
             this.file = file;
             this.logger.setFile(file);
 
+            // If file is not downloadable, trigger an error
+            if (file.is_download_available === false) {
+                const error = createPreviewError(ERROR_CODE.notDownloadable, __('error_not_downloadable'), null, {
```
[the remark this answers is not in the document]
Comment (JustinHoldstock) [reply src/lib/Preview.js]: The third param should be the details of the error. You can remove the third param, `null`. 

The second param of the function is set to the `displayMessage` property of the error created. The third is set to the `message` property of the error created.

Aww hell, I forgot that `null` wouldn't work with default params. I will make a ticket to update anywhere using `null`

File: src/lib/Preview.js [hunk 3]
```diff
@@ -1497,7 +1508,7 @@ class Preview extends EventEmitter {
                 console.error(message);
                 /* eslint-enable no-console */
 
-                const error = createPreviewError(ERROR_CODE, message, filesToPrefetch);
+                const error = createPreviewError(ERROR_CODE.prefetchFile, message, filesToPrefetch);
```
[the remark this answers is not in the document]
Comment (JustinHoldstock) [reply src/lib/Preview.js]: Yes. I apologize for missing that one. Thanks for catching it

File: src/lib/Preview.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (JustinHoldstock) [reply src/lib/Preview.js]: Just saw what you did there, with the fourth param. See comment below

File: src/lib/logUtils.js [hunk 4]
```diff
@@ -50,13 +50,20 @@ export function getClientLogDetails() {
...
     const error = new Error(code);
     error.code = code;
     error.message = details || displayMessage;
     error.displayMessage = displayMessage;
 
+    if (properties) {
+        Object.keys(properties).forEach((key) => {
+            error[key] = properties[key];
+        });
+    }
```
[the remark this answers is not in the document]
Comment (JustinHoldstock) [reply src/lib/logUtils.js]: I would say anything you want logged to the backend needs to be stored in `message`. If you want `linkUrl` and `linkText` to be sent with error details for debugging purposes, I'd agree that it makes sense to fold them into details. My only issue with that is we end up sending a translated string.

[the remark this answers is not in the document]
Comment (JustinHoldstock) [reply src/lib/logUtils.js]: One other note, I'm quite opposed to us dynamically creating properties on an already defined object type (`Error`). Now may be a good time for us to consider having our own class that extends `Error`, with our additional props on it, and *only* using that when creating errors in Preview.

File: src/lib/viewers/error/PreviewErrorViewer.js [hunk 5]
```diff
@@ -111,17 +117,31 @@ class PreviewErrorViewer extends BaseViewer {
...
+     *
+     * @param {string} linkText - Button message
+     * @param {string} linkUrl - URL for link
+     * @return {void}
+     */
+    addLinkButton(linkText, linkUrl) {
+        const linkBtnEl = this.infoEl.appendChild(document.createElement('a'));
+        linkBtnEl.className = 'bp-btn bp-btn-primary';
+        linkBtnEl.target = '_blank';
+        linkBtnEl.textContent = linkText;
```
Comment (pramodsum) [inline src/lib/viewers/error/PreviewErrorViewer.js]: Should this be localized too?

Comment (tonyjin) [reply src/lib/viewers/error/PreviewErrorViewer.js]: linkText will be localized before we pass it in.

File: src/lib/logUtils.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (tonyjin) [reply src/lib/logUtils.js]: @justinholdstock are you okay with my merging this as-is and then following up with a refactor? There are a lot of places I'd have to touch for a Previewer for refactor and 1) I'd like to get this change in before the release cut tomorrow and 2) I probably want that larger change in a separate PR.

[the remark this answers is not in the document]
Comment (JustinHoldstock) [reply src/lib/logUtils.js]: Absolutely! :D

**Pull request merged**